### PR TITLE
feat: Add `argocd_projects` field to Argocd addons

### LINKED ISSUE #12: Add support for `argocd_projects` field to Argocd addons
#### What is the outcome that you are trying to reach?
Request for a feature to create ArgoCD Projects for application isolation in ArgoCD. Currently, ArgoCD Addons support creating ArgoCD Applications, but not the ability to assign values for the project field from argocd_applications. The ability to create ArgoCD Projects would allow us to separate our applications based on environment, system, or workload.

#### Describe the solution you would like
Create an additional Helm chart for the ArgoCD Project (AppProject) and use a Terraform `helm_release` resource for `argocd_project`. Also, export the `argocd_projects` variables to create additional ArgoCD projects

### PR DESCRIPTION
**Pull request by dtherhtun**
### What does this PR do?

- Adding new `argocd_projects` field to ArgoCD addons. This feature allows users to create additional ArgoCD projects during the creation and update.

### Motivation

- Resolves #12 https://github.com/aws-ia/terraform-aws-eks-blueprints-addons/issues/56

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Yes, I ran `pre-commit run -a` with this PR

### For Moderators

- [ ] E2E Test successfully complete before merge?

### Additional Notes

I want to provide docs as well unfortunately there is no docs directory in this repo. But I wrote docs in this PR - https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/1413

### CODE REVIEW

File: modules/argocd/main.tf [hunk 1]
```diff
@@ -91,6 +91,46 @@ resource "helm_release" "argocd_application" {
   depends_on = [module.helm_addon]
```
Comment (csantanapr) [inline modules/argocd/main.tf:91]:
```suggestion
  depends_on = [module.argocd_project]
```
You need to create the projects first, if you have applications referring to one of the projects being created below.

Comment (dtherhtun) [reply modules/argocd/main.tf]: Thank you for pointing this out. updated the PR.

File: modules/argocd/main.tf [hunk 2]
```diff
@@ -96,6 +96,46 @@ resource "helm_release" "argocd_application" {
...
+# ---------------------------------------------------------------------------------------------------------------------
+# ArgoCD Projects Bootstrapping
+# ---------------------------------------------------------------------------------------------------------------------
+resource "helm_release" "argocd_project" {
+  for_each = { for k, v in var.projects : k => merge(local.default_argocd_project, v) }
+
+  name      = each.key
+  chart     = "${path.module}/argocd-project/helm"
+  version   = "1.0.0"
+  namespace = local.helm_config["namespace"]
```
Comment (csantanapr) [inline modules/argocd/main.tf:111]:
```suggestion
  namespace = lookup(each.value, "namespace", [])
```
allow to override the namespace from the map `projects`

File: modules/argocd/main.tf [hunk 3]
```diff
@@ -96,6 +96,46 @@ resource "helm_release" "argocd_application" {
...
+  values = [
+    # ArgoCD Project Spec
+    yamlencode({
+      "destinations"               = lookup(each.value, "destinations", [])
+      "clusterResourceWhitelist"   = lookup(each.value, "cluster_resource_whitelist", [])
+      "namespaceResourceBlacklist" = lookup(each.value, "namespace_resource_blacklist", [])
+      "namespaceResourceWhitelist" = lookup(each.value, "namespace_resource_whitelist", [])
+      "roles"                      = lookup(each.value, "roles", [])
+      "syncWindows"                = lookup(each.value, "sync_windows", [])
+      "sourceRepos"                = lookup(each.value, "repo_urls", [])
```
Comment (csantanapr) [inline modules/argocd/main.tf:135]:
```suggestion
      "sourceRepos"                = lookup(each.value, "source_repos", [])
```
consistency with key names

File: modules/argocd/locals.tf [hunk 4]
```diff
@@ -33,6 +33,29 @@ locals {
...
+    ]
+    namespace_resource_blacklist = []
+    namespace_resource_whitelist = []
+    destinations = [
+      {
+        namespace = "*"
+        server    = "https://kubernetes.default.svc"
+      }
+    ]
+    repo_urls    = ["*"]
```
Comment (csantanapr) [inline modules/argocd/locals.tf:54]:
```suggestion
    source_repos    = ["*"]
```

File: modules/argocd/argocd-project/helm/values.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#  - name: read-only
+#    description: Read-only privileges to my-project
+#    policies:
+#      - p, proj:my-project:read-only, applications, get, my-project/*, allow
+#    groups:
+#      - my-oidc-group
+# Destination cluster.
+destinations:
+  - namespace: "*"
+    server: "https://kubernetes.default.svc"
```
Comment (csantanapr) [inline modules/argocd/argocd-project/helm/values.yaml:32]:
```suggestion
    server: "https://kubernetes.default.svc"
sourceNamespaces: []
```

File: modules/argocd/argocd-project/helm/templates/project.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  namespaceResourceWhitelist:
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+
+  # Destination clusters can be identified by 'server', 'name', or both.
+  {{- with .Values.destinations }}
+  destinations:
+    {{- toYaml . | nindent 4  }}
+  {{- end }}
+
```
Comment (csantanapr) [inline modules/argocd/argocd-project/helm/templates/project.yaml:30]:
```suggestion
  {{- with .Values.sourceNamespaces }}
  sourceNamespaces:
    {{- toYaml . | nindent 4  }}
  {{- end }}
```

File: modules/argocd/locals.tf [hunk 7]
```diff
@@ -33,6 +33,29 @@ locals {
...
+      }
+    ]
+    namespace_resource_blacklist = []
+    namespace_resource_whitelist = []
+    destinations = [
+      {
+        namespace = "*"
+        server    = "https://kubernetes.default.svc"
+      }
+    ]
```
Comment (csantanapr) [inline modules/argocd/locals.tf:53]:
```suggestion
    ]
    source_namespaces = []
```

File: modules/argocd/main.tf [hunk 8]
```diff
@@ -96,6 +96,46 @@ resource "helm_release" "argocd_application" {
...
+
+  values = [
+    # ArgoCD Project Spec
+    yamlencode({
+      "destinations"               = lookup(each.value, "destinations", [])
+      "clusterResourceWhitelist"   = lookup(each.value, "cluster_resource_whitelist", [])
+      "namespaceResourceBlacklist" = lookup(each.value, "namespace_resource_blacklist", [])
+      "namespaceResourceWhitelist" = lookup(each.value, "namespace_resource_whitelist", [])
+      "roles"                      = lookup(each.value, "roles", [])
+      "syncWindows"                = lookup(each.value, "sync_windows", [])
```
Comment (csantanapr) [inline modules/argocd/main.tf:134]:
```suggestion
      "syncWindows"                = lookup(each.value, "sync_windows", [])
      "sourceNamespaces"           = lookup(each.value, "source_namespaces", [])
```

**Review by csantanapr — COMMENTED** [review]
add sourceNamespaces and rename variables
I'm using sourceNamespaces in this example https://github.com/csantanapr/terraform-aws-eks-blueprints/blob/argo-multi-cluster/examples/gitops/argocd-multi-cluster/spoke-cluster-template/main.tf#L281

File: modules/argocd/main.tf (hunk 2 above)
Comment (dtherhtun) [reply modules/argocd/main.tf]: updated but it is still using `namespace = local.helm_config["namespace"]` in `applications`.

File: modules/argocd/main.tf (hunk 3 above)
Comment (dtherhtun) [reply modules/argocd/main.tf]: updated.

File: modules/argocd/locals.tf (hunk 4 above)
Comment (dtherhtun) [reply modules/argocd/locals.tf]: updated

File: modules/argocd/argocd-project/helm/values.yaml (hunk 5 above)
Comment (dtherhtun) [reply modules/argocd/argocd-project/helm/values.yaml:32]: I don't think we need `sourceNamespaces: []`. actually, `destinations:` is `destinations: []` and we can add more namespace and server in the `destinations`. please take a look at this example - https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/project.yaml . I'm referencing it.

Comment (dtherhtun) [reply modules/argocd/argocd-project/helm/values.yaml:32]: here is my example - https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/1413/files#diff-35e3176ae40868f713cd61ae43525df41808209c100b21562ffaf66e699bc42d

Comment (csantanapr) [reply modules/argocd/argocd-project/helm/values.yaml:32]: Hi @dtherhtun 
destinations are different sourceNamespaces
Here are the docs https://argo-cd.readthedocs.io/en/stable/operator-manual/app-any-namespace/#managing-applications-in-other-namespaces

This is used when you have ArgoCD applications in different namespaces other than `argocd` namespace that the application controller is installed.
The use case is when a single ArgoCD instance is managing different remote clusters, and you want to create namespaces for tenants to install ArgoCD Applications in different namespaces with the same name (ie addon name), but for different clusters.
See my example here:
- https://github.com/csantanapr/terraform-aws-eks-blueprints/blob/argo-multi-cluster/examples/gitops/argocd-multi-cluster/hub-cluster/helm-argocd/values.yaml#L85
- https://github.com/csantanapr/terraform-aws-eks-blueprints/blob/argo-multi-cluster/examples/gitops/argocd-multi-cluster/spoke-cluster-template/main.tf#L353

File: modules/argocd/main.tf (hunk 2 above)
Comment (csantanapr) [reply modules/argocd/main.tf]: Yes this is intentional, we want to allow the flexibility to install application and projects in different namespaces, and this way if user doesn't specify a different namespace via project.namepspace it defaults to `local.helm_config["namespace"]` via:
```
default_argocd_project = {
    namespace   = local.helm_config["namespace"]
```

File: modules/argocd/argocd-project/helm/values.yaml (hunk 5 above)
Comment (dtherhtun) [reply modules/argocd/argocd-project/helm/values.yaml:32]: @csantanapr updated and thanks.

**Comment by dtherhtun** [comment]
@bryantbiggs why is this PR only left always? what do I need to do?

**Comment by bryantbiggs** [comment]
I will defer to @csantanapr for review - I was simply updating the PR to see if we were able to resolve the pr-title check issue (which it seems we have not, yet. not related to this PR)

**Review by csantanapr — APPROVED** [review]
thank you for addressing the changes

**Pull request merged**